R/method.R

```diff
@@ -1,145 +1,163 @@
+  invisible(generic)
 }
 
-methods <- function(generic) {
-  get_all_methods(generic@methods, character())
-}
+register_method <- function(generic, signature, method, package = packageName(parent.frame())) {
+  signature <- as_signature(signature)
+  generic <- as_generic(generic)
 
-get_all_methods <- function(x, signature) {
-  if(!is.environment(x)) {
-    return(x)
+  if (is_external_generic(generic)) {
+    if (!is.null(package)) {
+      # method registration within package, so add to lazy registry
+      external_methods_add(package, generic, signature, method)
+    } else {
+      # otherwise find the generic and register
+      generic <- getFromNamespace(generic$name, asNamespace(generic$package))
+      register_method(generic, signature, method)
+    }
+  } else if (is_s3_generic(generic)) {
+    generic_name <- attr(generic, "name")
+
+    if (length(signature) != 1 || class_type(signature[[1]]) != "r7") {
+      msg <- sprintf(
+        "When registering methods for S3 generic %s(), signature be a single R7 class",
+        generic_name
+      )
+      stop(msg, call. = FALSE)
+    }
+    class <- signature[[1]]@name
+    registerS3method(generic_name, class, method, envir = parent.frame())
+  } else {
+    check_method(method, signature, generic)
+    register_r7_method(generic, signature, method)
   }
 
-  unlist(lapply(names(x), function(class) get_all_methods(x[[class]], c(signature, class))), recursive = FALSE)
+  invisible()
 }
 
-as_signature <- function(signature) {
-  if (!is.list(signature)) {
-    signature <- list(signature)
-  }
+register_r7_method <- function(generic, signature, method) {
+  # Flatten out unions to individual signatures
+  signatures <- flatten_signature(signature)
 
-  for (i in seq_along(signature)) {
-    signature[[i]] <- as_class(signature[[i]], arg = sprintf("signature[[%i]]", i))
+  # Register each method
+  for (signature in signatures) {
+    method <- R7_method(method, generic = generic, signature = signature)
+    generic_add_method(generic, signature, method)
   }
-  signature
+
+  invisible()
 }
 
-method_impl <- function(generic, signature, ignore) {
-  out <- .Call(method_, generic, signature, ignore)
-  if (!is.null(out)) {
-    return(out)
-  }
+flatten_signature <- function(signature) {
+  # Unpack unions
+  sig_is_union <- vlapply(signature, is_union)
+  signature[sig_is_union] <- lapply(signature[sig_is_union], prop, "classes")
+  signature[!sig_is_union] <- lapply(signature[!sig_is_union], list)
 
-  # If no R7 method is found, see if there are any S3 methods registered
-  if (inherits(generic, "R7_generic")) {
-    args <- generic@dispatch_args
-    generic <- generic@name
-  } else {
-    args <- setdiff(names(formals(generic)), "...")
-    generic <- find_function_name(generic, topenv(environment(generic)))
-  }
+  # Create grid of indices
+  indx <- lapply(signature, seq_along)
+  comb <- as.matrix(rev(do.call("expand.grid", rev(indx))))
```

Comment:
Inlining a base R version of `tidyr::expand_grid()` to use on `indx` would also be pretty straightforward for this.

If you keep it this way, it might help to add a comment stating that the point is to vary the first column the slowest, unlike `expand.grid()`